### ISSUE TITLE
How to restrict extensionDependencies version

### ISSUE THREAD

**Issue by dilin-MS** [microsoft/vscode#82655]
Hi, I would like to know how to restrict extensionDependencies version?
I have a main extension A (Azure IoT Device Workbench), which has extension B (IoT Device Cube) as its dependent extension.
In extension A's package.json:
```
  "extensionDependencies": [
    "vsciot-vscode.vscode-iot-device-cube"
  ]
```

At first, extension A `v1.2.3` will install latest `v0.0.1` extension B. Later extension B has **breaking change** and publish `v0.2.0`. To fix incomapability with the latest extension B, extension A publish a new release `v1.2.4` which works well with extension B `v0.2.0`. Problem is that user installing extension A old release `v1.2.3` will also install latest extension B `v0.2.0`, this is wrong. 

How can I restrict extension Dependencies version so extension A `v1.2.3` can use `extensionB: ^0.0.1`. while extension A `v1.2.4` can use `extensionB: ^0.2.0`.

**Comment by sandy081** [comment]
Currently there is no such possibility. Extension dependencies are always expected to be latest.

**Comment by alexandrudima** [comment]
@dilin-MS Currently, a certain extension can only execute once in VS Code. It cannot execute at two distinct versions at the same time. This is due to contributions / API usage which might lead to duplicates or overriding behavior (imagine an extension that contributes a status bar entry, that status bar entry would appear two times, etc.).

We have benefited immensely from VS Code updates. We always ask people to update before we attempt to reproduce an issue... We want that same experience for our extensions. We don't want for an extension author to be forced to maintain or support an older version of their extension.

That being said, the pattern we have used successfully is something like the following:

From the API provider - `extensionA` - export a single method called `export function getAPI(version: number): API`. Whenever it does a breaking change, `extensionA` needs to continue returning valid results for older version numbers. `extensionA` can choose for how long it wants to support its old versions.

**Comment by dilin-MS** [comment]
From your reply, you suggest we would like user to use the latest extension. But there are situations that user need to use old version of extension. It's just like node modules which we would have `^` `~` symbol to restric the version of dependency. **It is very passive if we could not restric the dependency version we are using.** Since every breaking change of our dependency may cause us to break.

From my scenario, `v1.2.3` and `v1.2.4` of extension A mainly differs on the **VS Code engine requirement**. If user has lower version VS Code engine, he cannot install the latest extension A `v1.2.4`. Instead, he will fall back to `v1.2.3` because that's the usable version for his VS Code engine.
In this scenario, we would like to ensure both `v1.2.3` and `v1.2.4` of extension A is usable. 

Now the problem is how to ensure `v1.2.3` extension A workable, leting it installs `v0.0.1` extension B instead of `v2.0.0`.

In this scenario, what's the recommendation from vscode's design philosophy?

**Comment by alexandrudima** [comment]
@dilin-MS 
I'm not 100% sure. Please help me better understand your use-case and your constraints.

I have looked at https://github.com/VSChina/vscode-iot-device-cube/blob/master/src/extension.ts and from what I can tell that extension registers commands, which then I think are being invoked by other extensions.

Is there any reason why https://github.com/VSChina/vscode-iot-device-cube is not published as a node module and consumed via npm?

**Comment by dilin-MS** [comment]
Yes, `vsciot-vscode.vscode-iot-device-cube` is a helper extension of `vsciot-vscode.vscode-iot-workbench`([Azure IoT Device Workbench extension](https://marketplace.visualstudio.com/items?itemName=vsciot-vscode.vscode-iot-workbench)).
The helper cube extension is an [UI extension](https://code.visualstudio.com/api/advanced-topics/remote-extensions#architecture-and-extension-kinds) exists to support interaction with local file system when workbench extension works in remote environment.

**Comment by alexandrudima** [comment]
Thanks for the extra information, that is helpful.

I personally like to think about this in a similar way a REST API works. You can think that the `vsciot-vscode.vscode-iot-device-cube` is the "server"/"provider" and `vsciot-vscode.vscode-iot-workbench` is the "client"/"consumer". Usually, this problem appears in client-server scenarios, when you have a server that might update to version N+1 and you still want to support clients at version N... Sometimes, it is possible to have backwards compatibility for even more than one version.

I would do the following:
1. change all commands to take as argument a single object. Instead of `fsCopyFile(sourcePath: string, targetPath: string)`, I would do `fsCopyFile(opts: { sourcePath: string; targetPath: string})`. Changing the arguments from an array to an object would be helpful because it would allow you to evolve the options, to add new fields which you haven't considered yet.
2. add a version property to each and every command. So that would make it look like `fsCopyFile(opts: { version: number; sourcePath: string; targetPath: string})`. You can then specify something like the following in TS:
```ts
interface ICopyFileV1 { version: 1; sourcePath: string; targetPath: string; }
// suppose you want to change something, like rename properties/add properties
interface ICopyFileV2 { version: 2; source: string; target: string; overwrite: boolean }
type ICopyFileResult = boolean;

// suppose you want to support both versions 1 and 2
async function fsCopyFile(opts: ICopyFileV1 | ICopyFileV2): Promise<ICopyFileResult> {
  // you can now check if opts.version is 1 or 2 and provide
  // backwards compatibility for 1, with proper TS compiler typing
}
```
3. Distribute to your clients a versioned API via a `.d.ts` file that is always the latest:
```ts
interface IAPI {
  ('iotcube.fsCopyFile', opts: { version: 2; source: string; target: string; overwrite: boolean }): Promise<boolean>;
}

const iotcubeAPI = <IAPI>(vscode.commands.executeCommand);
```
4. Once your clients are on the latest API, you can potentially clean up and remove support for old versions.

**Comment by jdneo** [comment]
The Java side has the same request. For example:

The Java Language Server extension supports a new capability since version `A`. As its dependent extension -- Java Test Runner, it wants to integrate the new capability and provide a new feature.

Currently our approach is to get the version inside the extension's `package.json`, and use npm modules like `version-compare` to check if the new capability is available or not. 

If the VS Code can officially support the extension dependency version, it will be much more easier for the extension author. 😄 

// cc @testforstephen if more examples want to be added.

**Comment by andyliuliming** [comment]
yeah. I think stick to one version is critical for one mature software system.
breaking customer with a random time is not good.

**Comment by zhongyi-zhang** [comment]
Same request here. I think this is a basic feature. If we can't restrict dependency version, maybe a new installation of my extension still works today and just broke tomorrow.
